src/plugins/grpc/server.js

```diff
@@ -0,0 +1,137 @@
+'use strict'
+
+const { FORMAT_HTTP_HEADERS, Tags } = require('opentracing')
+
+let patchedHandlers = []
+
+function createWrapHandler (grpc, tracer, config, handler) {
+  return function wrapHandler (func) {
+    return function funcHandler (emitter) {
+      const { metadata, request } = emitter
+      const { type } = this
+      const isStream = ['client_stream', 'server_stream'].includes(type)
+      const childOf = tracer.extract(FORMAT_HTTP_HEADERS, metadata.getMap())
+      const span = tracer.startSpan('grpc.server', {
+        childOf,
+        tags: {
+          [Tags.SPAN_KIND]: Tags.SPAN_KIND_RPC_SERVER,
+          'grpc.method_name': handler,
+          'grpc.method_type': type,
+          'resource.name': handler.split('/').slice(-1),
+          'service.name': config.service || `${tracer._service}-grpc`
+        }
+      })
+
+      tracer.scopeManager().activate(span)
+
+      if (request) {
+        if (metadata) {
+          for (const [key, value] of Object.entries(metadata.getMap())) {
+            span.setTag(`grpc.metadata.${key}`, value)
+          }
+        }
+
+        if (config.arguments) {
+          const values = config.arguments(request)
+
+          for (const key in values) {
+            span.setTag(`grpc.arguments.${key}`, values[key])
+          }
+        } else {
+          span.setTag('grpc.arguments', request)
+        }
+      }
+
+      if (isStream) {
+        emitter.on('error', err => {
+          handleError(config, span, err)
+
+          span.finish()
+        })
+
+        // Finish the span if the response only if it was successful.
+        // Otherwise it'll be finished in the `error` listener.
+        emitter.on('finish', function () {
+          if (emitter.status.code === 0) {
```

Comment:
Is it possible that the status is not `0` and the `error` event is not emitted?

Reply:
Also, what happens if the request is cancelled?

Reply:
> Is it possible that the status is not 0 and the error event is not emitted?

Hard to say but from looking at the [server implementation](https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/server.js) I'd consider it a bug if that's possible.

> Also, what happens if the request is cancelled?

Unfortunately I can't seem to find a way to support cancelled requests. A cancelled unary call will [never call](https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/server.js#L587-L589) our wrapped function.

It would be great if we could patch the [call.startBatch()](https://github.com/grpc/grpc-node/blob/b1e7d446265065f35471dc533e25db5a8ab4f5d9/packages/grpc-native-core/src/server.js#L575) or the [handleUnaryCall()](https://github.com/grpc/grpc-node/blob/b1e7d446265065f35471dc533e25db5a8ab4f5d9/packages/grpc-native-core/src/server.js#L567) calls but I just couldn't find an injection point to patch them. Totally open to suggestions if you have an idea on this one.

Reply:
This is unfortunately a blocker, since not finishing a span will result in a memory leak. I would have to take a look, but hopefully there is a construct available to detect that.

Is the emitter also an actual stream? If that's the case maybe it emits some of the standard events on cancellation?

Reply:
A test for this is definitely needed as well.

Reply:
Makes sense. I'll see if I can come up with a way to handle cancellations. Will get back to you soon.